packages/go/analysis/ad/adcscache.go

```diff
@@ -91,51 +99,125 @@ func (s ADCSCache) BuildCache(ctx context.Context, db graph.Database, enterprise
-				log.Warnf("Error checking HasWeakCertBindingInForest for domain %d: %v", domain.ID, err)
+			if weakCertBinding, err := hasWeakCertBindingInForest(tx, domain); err != nil {
+				log.Warnf("Error checking hasWeakCertBindingInForest for domain %d: %v", domain.ID, err)
 				return nil
 			} else if weakCertBinding {
-				s.HasWeakCertBindingInForest[domain.ID] = struct{}{}
+				s.hasWeakCertBindingInForest.Add(domain.ID.Uint32())
 			}
 		}
 
 		return nil
 	})
+	if err != nil {
+		log.Errorf("Error building adcs cache %v", err)
+	}
 
 	log.Infof("Finished building adcs cache")
 }
 
-func (s ADCSCache) DoesCAChainProperlyToDomain(enterpriseCA, domain *graph.Node) bool {
+func (s *ADCSCache) DoesCAChainProperlyToDomain(enterpriseCA, domain *graph.Node) bool {
 	var domainID = domain.ID
 	var caID = enterpriseCA.ID.Uint32()
 
-	if _, ok := s.RootCAForChainValid[domainID]; !ok {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	if _, ok := s.rootCAForChainValid[domainID]; !ok {
 		return false
-	} else if _, ok := s.AuthStoreForChainValid[domainID]; !ok {
+	} else if _, ok := s.authStoreForChainValid[domainID]; !ok {
 		return false
 	} else {
-		return s.RootCAForChainValid[domainID].Contains(caID) && s.AuthStoreForChainValid[domainID].Contains(caID)
+		return s.rootCAForChainValid[domainID].Contains(caID) && s.authStoreForChainValid[domainID].Contains(caID)
 	}
 }
 
-func HasUPNCertMappingInForest(tx graph.Transaction, domain *graph.Node) (bool, error) {
+func (s *ADCSCache) GetExpandedCertTemplateControllers(id graph.ID) cardinality.Duplex[uint32] {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if expandedCertTemplateControllers, ok := s.expandedCertTemplateControllers[id]; !ok {
+		return cardinality.NewBitmap32()
+	} else {
+		return expandedCertTemplateControllers
+	}
+}
+
+func (s *ADCSCache) SetExpandedCertTemplateControllers(certId graph.ID, principalId uint32) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if _, ok := s.expandedCertTemplateControllers[certId]; !ok {
+		s.expandedCertTemplateControllers[certId] = cardinality.NewBitmap32With(principalId)
+	} else {
+		s.expandedCertTemplateControllers[certId].Add(principalId)
+	}
+}
+
+func (s *ADCSCache) GetCertTemplateEnrollers(id graph.ID) []*graph.Node {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.certTemplateEnrollers[id]
+}
+
+func (s *ADCSCache) GetCertTemplateControllers(id graph.ID) []*graph.Node {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.certTemplateControllers[id]
+}
+
+func (s *ADCSCache) GetEnterpriseCAEnrollers(id graph.ID) []*graph.Node {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.enterpriseCAEnrollers[id]
+}
+
+func (s *ADCSCache) GetPublishedTemplateCache(id graph.ID) []*graph.Node {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	published := s.publishedTemplateCache[id]
+	return published
+}
+
+func (s *ADCSCache) HasUPNCertMappingInForest(id uint32) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	hasUPNCertMappingInForest := s.hasUPNCertMappingInForest.Contains(id)
```

Comment:
Probably could just be collapsed into the return statement like the functions above.